### PR TITLE
Add INDEX USING HASH support

### PR DESCRIPTION
**Pull request by de-luca**
Hi,

This PR adds support for [hash sharded indexes](https://www.cockroachlabs.com/docs/v23.1/hash-sharded-indexes).

It adds `USING HASH [WITH ( bucket_size = X )]` to:
- column constraints
- table constraints
- CREATE INDEX statements

Also added:
- support for alternate `STORING` keyword: `INCLUDE` and `COVERING`. 
- `CONCURRENTLY` on `CREATE INDEX`

### CODE REVIEW

File: cockroachdb-dialect/src/main/kotlin/com/faire/sqldelight/dialects/cockroachdb/grammar/CockroachDB.bnf [hunk 1]
```diff
@@ -134,10 +156,10 @@ blob_data_type ::= 'BYTEA' | 'BLOB' | 'BYTES' {
   override = true
 }
 
-create_index_stmt ::= CREATE [ UNIQUE ] INDEX [ IF NOT EXISTS ] [ [ ansi_database_name DOT ] ansi_index_name ] ON ansi_table_name LP ansi_indexed_column ( COMMA ansi_indexed_column ) * RP ['STORING' LP ansi_indexed_column ( COMMA ansi_indexed_column ) * RP ] [ WHERE <<expr '-1'>> ] {
+create_index_stmt ::= CREATE [ UNIQUE ] INDEX [ 'CONCURRENTLY' ] [ IF NOT EXISTS ] [ [ ansi_database_name DOT ] ansi_index_name ] ON ansi_table_name LP ansi_indexed_column ( COMMA ansi_indexed_column ) * RP [ index_using_hash ] [('STORING' | 'COVERING' | 'INCLUDE') LP ansi_indexed_column ( COMMA ansi_indexed_column ) * RP ] [ WHERE <<expr '-1'>> ] {
```
Comment (Adriel-M) [inline cockroachdb-dialect/src/main/kotlin/com/faire/sqldelight/dialects/cockroachdb/grammar/CockroachDB.bnf:159]: Can you add a test case for `CONCURRENTLY`

File: cockroachdb-dialect/src/main/kotlin/com/faire/sqldelight/dialects/cockroachdb/grammar/CockroachDB.bnf [hunk 2]
```diff
@@ -94,8 +99,10 @@ overrides ::= table_constraint
   | alter_table_rules
   | extension_stmt
 
+index_using_hash ::= 'USING' 'HASH' [ storage_options ]
```
Comment (Adriel-M) [inline cockroachdb-dialect/src/main/kotlin/com/faire/sqldelight/dialects/cockroachdb/grammar/CockroachDB.bnf:102]: Might be good to use something else than storage_options and create your own just for the hash options.

The cockroach grammar only supports one option https://www.cockroachlabs.com/docs/stable/sql-grammar#opt_hash_sharded

While storage options can support multiple.

Comment (Adriel-M) [reply cockroachdb-dialect/src/main/kotlin/com/faire/sqldelight/dialects/cockroachdb/grammar/CockroachDB.bnf:102]: Also there should be a SqlType for `USING` already

**Review by Adriel-M — COMMENTED** [review]
Can you also add a migration file for the integration test for the index hashing support? No need to assert anything. Just need to make sure it actually runs.

Folder can be found: https://github.com/Faire/sqldelight-cockroachdb-dialect/tree/main/integration-testing/src/main/resources/migration/com/faire/sqldelight/dialects/cockroachdb

You can run the gradle task `:integration-testing:generateMigrationFile` to create a migration file.

Thanks!

File: cockroachdb-dialect/src/main/kotlin/com/faire/sqldelight/dialects/cockroachdb/grammar/CockroachDB.bnf (hunk 1 above)
Comment (de-luca) [reply cockroachdb-dialect/src/main/kotlin/com/faire/sqldelight/dialects/cockroachdb/grammar/CockroachDB.bnf:159]: Yes, I forgot that one, my bad

**Review by Adriel-M — APPROVED** [review]
Thanks a lot!

**Pull request merged**